New feature/vignette project (#108)
dfe-analytical-services/dfeR: vignettes/create_new_project.Rmd

## Using the `create_project()` function

The `create_project()` function is a part of the dfeR package and is used to create a new R project with a custom folder structure. It is designed to help you set up your project quickly and efficiently, with all the necessary components in place. 

The function works by:

1.  **Setting up the core project structure** The core project structure will either create an ad-hoc project folder structure, which includes folders for data, analysis and output, or a larger folder structure for publication projects, which includes more folders for different data, analysis, documentation and outputs.

2.  **Adding initial files** The script creates some initial files, such as a template README file to help you get started with your project. It also initialises the `testthat` package, which is used for unit testing in R.

> 3.  **Customisation** The project creates a GitHub .gitignore file if set to `TRUE` and initialises the `renv` package if set to `TRUE`.

Author: Looks good thanks, I had looked at this sentence and gone back and forth on changing as wasnt sure it was clear, particularly with the 'and' feeling a bit weird!

Here is the basic code you will need to use, having installed dfeR:

```{r create_project, eval = FALSE}
dfeR::create_project(
  path = "C:/path/to/your/new/project",
  init_renv = TRUE,
  include_structure_for_pkg = FALSE,
  create_publication_proj = FALSE,
  include_github_gitignore = TRUE
)
```

### Where to save your project 

You should store any projects that use Git outside of OneDrive. To create a folder in your C drive, type: 'C:/Users/username' into the file explorer search bar and create a folder called 'repos'.

### Parameters

Below is a table describing the parameters used within the code. When using the create_project function, you will need to make a decision on what to enter for each in order to create the appropriate type of project.

```{r, echo = FALSE}
parameter_data <- data.frame(
  Parameter = c(
    "Path",
    "init_renv",
    "include_structure_for_pkg",
    "create_publication_proj",
    "include_github_gitignore"
  ),
  Meaning = c(
    "Folder pathway",
    "Initiate renv in the project",
    "Additional folder for package development",
    "Create project with structure for publications",

>     "Include a github .gitignore file"

Reviewer: ```suggestion
    "Include a GitHub .gitignore file"
```

  ),
  `Expected input` = c(
    "A folder path e.g. 'C:/Users/username/repos'",
    "Boolean (TRUE/FALSE), default = TRUE",
    "Boolean (TRUE/FALSE), default = FALSE",
    "Boolean (TRUE/FALSE), default = FALSE",
    "Boolean (TRUE/FALSE)"
  ),
  Considerations = c(
    "Where do you want to store your project?",
    "Do you want to use renv for package version control?",
    "Are you creating this project to work on a package?",
    "Should the folder structure be for a publication project?",
    "Do you want to exclude certain files from being tracked by Git?"
  ),
  Output = c(
    "This will be the folder your new project is created within.",
    "If set to TRUE, renv will be initialised in your project.",
    "If set to TRUE, an additional package development folder will be created",
    "If set to TRUE, extra publication specific folders will be created",

>     "If set to TRUE, a github .gitignore file will be created."

Reviewer: ```suggestion
    "If set to TRUE, a GitHub .gitignore file will be created."
```

)
```

You will also need to think about what you want your project to be called and include this in your file path. You can always amend this later.

For example, if you want to create a project called “my-new-project” in the folder you created above, your file path (with the username changed to your username) would be "C:/Users/Msmith/repos/my-new-project".

You can create a project in a different location, but you will need to change the file path accordingly.

### Example

```{r function_example, eval = FALSE}
# Load the dfeR package
library(dfeR)
# Create a new project with the desired parameters
dfeR::create_project(
  path = "C:/Users/MSmith/repos/my-new-project",
  init_renv = TRUE,
  include_structure_for_pkg = FALSE,
  create_publication_proj = TRUE,
  include_github_gitignore = TRUE
)
```

The above code will create a new project in the specified folder path, with renv initialised, a publication project structure, and a GitHub .gitignore file included. You can view this by going to the location of the pathway you chose in file explorer. You should find your new project there, inside a folder with the name you added at the end of your pathway, as below.

> ![Newly created contents of project folder in file explorer](create_new_project_screenshot.png)

Author: Yes you're right sorry, that should be alt text! Was just the way I put it in this time - have changed :)

### What is an R Project?

An R project is a working directory designated for a specific analysis or project. It keeps all your scripts, data, functions and outputs together in one place. Rstudio allows you to switch between your projects easily, meaning you can keep your different pieces of work separate from each other.

Whenever you are using R studio, you should work in an R Project. This makes sure you are set up in the correct working directory, so your code is pointing at the right folders and files. You can find out more in [this guide for using projects in R](https://support.posit.co/hc/en-us/articles/200526207-Using-RStudio-Projects).

### What is renv?

Packages and versions of R regularly update. Over time, this can cause code to break - e.g. if different dependencies are required for later versions of packages to work. Using renv creates a “snapshot” of your code and packages at the time you created it, which anyone can then recreate when they come to use your code. This means that you can be sure that the code you wrote will work in the future, even if packages are updated.

> You should use the renv package for package version control in R. It creates a project-specific library, meaning that installing a new or updated package for one project won't break your other projects. This is particularly useful for reproducibility and collaboration, as it ensures that your project will work with the same package versions on different machines. You can find out more about renv on the the [RStudio website](https://rstudio.r-universe.dev/renv#:~:text=A%20dependency%20management%20toolkit%20for%20R.%20Using%20%27renv%27%2C,make%20your%20projects%20more%20isolated%2C%20portable%2C%20and%20reproducible._).

Author: Ahh this is due to the lintr checks meaning I had to add a line break!

    "Boolean (TRUE/FALSE), default = FALSE",
    "Boolean (TRUE/FALSE)"
  ),
  Considerations = c(
    "Where do you want to store your project?",
    "Do you want to use renv for package version control? 
    Set to TRUE if you want to use it.",
    "Are you creating this project to work on a package?
    Set to TRUE if yes.",
    "Should the folder structure be for a publication project?
    Set to TRUE if yes or FALSE if it is for another type of analysis.",
    "Do you want to exclude certain files from being tracked by Git?"
  ),
  Outout = c(
    "This will be the folder your new project is created within",
    "If set to TRUE, renv will be initialised in your project.
    A renv folder will be created, as well as a renv.lock file.",
    "If set to TRUE, an additonal folder will be created specifically
    for package development",
    "If set to TRUE, a publication template folder structure will be created,
    with folders for data, analysis, documentation and misc. 
    If set to FALSE, a basic analysis folder structure will be created,
    with folders for data, analyis and output",
    "If set to TRUE, a github gitignore file will be created."
  )
)

> kable(

Reviewer: I think something went wrong here. this is what it looks like when i render: 
![image](https://github.com/user-attachments/assets/990832b6-6050-443c-b7bd-90ac5b49056f)

### Example

```{r function_example, eval = FALSE}
# Load the dfeR package
library(dfeR)
# Create a new project with the desired parameters
dfeR::create_project(
  path = "C:/Users/MSmith/repos/my-new-project",
  init_renv = TRUE,
  include_structure_for_pkg = FALSE,
  create_publication_proj = TRUE,
  include_github_gitignore = TRUE
)
```

The above code will create a new project in the specified folder path, with renv initialised, a publication project structure, and a GitHub gitignore file included. You can view this by going to the location of the pathway you chose in file explorer. You should find your new project there, inside a folder with the name you added at the end of your pathway, as below.

![Newly created contents of project folder in file explorer](create_new_project_screenshot.png)

You can then open the project by going to file \> open project in RStudio. If you have set create_publication_proj to FALSE, you will find your folders will look a little different to the above.

You can amend the file structure and names to suit the particular project you are working on, using the file structure created as a template.

### Troubleshooting

> You may hit an error if the `init_renv` parameter is not set to FALSE. This initiates `{renv}` in the created project, which is best practice. However, the `renv.lock` is set up to record several packages (`{rmarkdown}, {testthat}, {renv}, {stringr}`). You may get the following error message, letting you know that none of these packages are installed:

Reviewer: ```suggestion
You may hit an error if the `init_renv` parameter is not set to FALSE. This initiates `renv` in the created project, which is best practice. However, the `renv.lock` is set up to record several packages (`rmarkdown, testthat, renv, stringr`). You may get the following error message, letting you know that none of these packages are installed:
```

### What is testthat?

`testthat` is an R package that helps you write and run tests to ensure your code works correctly. It allows you to create tests that check if your functions produce the expected results, run these tests to catch errors early, and gives you confidence that your code is reliable and behaves as intended. It's a valuable tool for maintaining high-quality, bug-free code.

See this [guide on testing](https://r-pkgs.org/testing-basics.html) by Hadley Wickham for getting started.

### What is a GitHub .gitignore file?

A .gitignore file is a text file that tells Git which files or directories to ignore in a project. It helps you avoid committing sensitive information, temporary files, or other unnecessary files to your Git repository. By creating a .gitignore file, you can keep your repository clean and focused on the essential files needed for your project. Find more guidance on how to avoid revealing sensitive information on the [Analysts' Guide](https://dfe-analytical-services.github.io/analysts-guide/RAP/rap-statistics.html#avoid-revealing-sensitive-information).

> ## Using the create_project() function {#using-the-create_project-function}

Reviewer 2: Nice - haven't seen setting the custom URL like this before!

The `create_project()` function is a part of the `dfeR` package and is used to create a new R project with a custom folder structure. It is designed to help you set up your project quickly and efficiently, with all the necessary components in place.

The function works by:

1.  **Setting up the core project structure**: The core project structure will either create an ad-hoc project folder structure, which includes folders for data, analysis and output, or a larger folder structure for publication projects, which includes more folders for different data, analysis, documentation and outputs.

2.  **Adding initial files**: The script creates some initial files, such as a template README file to help you get started with your project. It also initialises the `testthat` package, which is used for unit testing in R.

3.  **Customisation**: Other files or processes can also be initialised depending on whether you set extra arguments as `TRUE` or `FALSE`. For example, you can use customisation to create a GitHub .gitignore file or initialise the `renv` package.

### Where to save your project

> You should store any projects that use Git outside of OneDrive. To create a folder in your C drive, type: 'C:/Users/username' into the File Explorer search bar and create a folder called 'repos'.

Reviewer 2: Suggest using an example to make this really clear

```suggestion
You should store any projects that use Git outside of OneDrive. To create a folder in your C drive, type: 'C:/Users/[username]' into the File Explorer search bar and create a folder called 'repos'. For example, if your user was JBLOGGS, you'd type 'C:/Users/JBLOGGS'.
```

### Parameters

When creating a project using the `create_project()` function, you will need to decide on several parameters. These parameters are necessary whether you use code or the 'New Project Wizard' to create your project. If you are using the 'New Project Wizard', the parameters which need a `TRUE`/`FALSE` input will be set to `TRUE` by ticking a checkbox. The table below explains each of the parameters, their meaning and what you should take into account for each.

> ```{r, echo = FALSE, warning=FALSE, message=FALSE}

Reviewer 2: Neat use of kable / R to make the table!

![image](https://github.com/user-attachments/assets/c1404b4a-2fdf-4a36-95d1-d87bdb65ca03)
